File: Sort/QuickSort.cpp
#include <iostream>
#include <algorithm>

int partition(float A[], int p, int r){
    int i = p , j = r+1;
    float x = A[p];
    while(1){
        while(A[++i] < x);
        while(A[--j] > x);
        if(i >= j)break;
        std::swap(A[i], A[j]);
    }
    A[p] = A[j];
    A[j] = x;
    return j;
}

void QuickSort(float a[], int p, int r){
    if(p < r){
        int q = partition(a, p, r);
        QuickSort(a, p, q-1);
        QuickSort(a, q+1, r);
    }
}

int main(){
    float A[] = {23, 4, 6, 7, 8, 2, 5, 9, 33};
    QuickSort(A, 0, 8);
    for(int i = 0; i <= 8; i++){
        std::cout<<A[i]<<" ";
    }
}

File: Sort/QuickSort.go
package main

import "fmt"

func partition(a []int, p int, r int)int {
	i, j := p, r+1
	x := a[p]
	i = i + 1
	j = j + 1
	for true {
		for a[i] < x {i++}

		for a[j] > x {j++}

		if i >= j {
			break
		}

		temp := a[i]
		a[i] = a[j]
		a[j] = temp
	}
	a[p] = a[j]
	a[j] = x
	return j


}

func QuickSort(a []int, p int, r int){
	if p < r {
		q := partition(a, p, r)
		QuickSort(a, p, q-1)
		QuickSort(a, q+1, r)
	}
}

func main()  {
	a := []int{3, 1, 2, 5, 6, 43, 4}
	QuickSort(a, 0, len(a)-1)
	fmt.Println(a)
}